File: SparkStreamingAction/spark-streaming-train/src/main/scala/com/yourtion/bigdata/spark/project/dao/CourseClickCountDAO.scala
package com.yourtion.bigdata.spark.project.dao

import com.yourtion.bigdata.spark.project.domain.CourseClickCount
import com.yourtion.bigdata.spark.project.utils.HBaseUtils
import org.apache.hadoop.hbase.client.Get
import org.apache.hadoop.hbase.util.Bytes

import scala.collection.mutable.ListBuffer

/**
 * 实战课程点击数-数据访问层
 */
object CourseClickCountDAO {

  val tableName = "imooc_course_click_count"
  val cf = "info"
  val qualifier = "click_count"

  def main(args: Array[String]): Unit = {
    val list = new ListBuffer[CourseClickCount]
    list.append(CourseClickCount("20200101_8", 8))
    list.append(CourseClickCount("20200101_9", 9))
    list.append(CourseClickCount("20200101_1", 100))

    save(list)

    println(count("20200101_8") + " : " + count("20200101_9") + " : " + count("20200101_1"))
  }

  /**
   * 保存数据到HBase
   *
   * @param list CourseClickCount集合
   */
  def save(list: ListBuffer[CourseClickCount]): Unit = {
    val table = HBaseUtils.getInstance().getTable(tableName)

    for (ele <- list) {
      table.incrementColumnValue(Bytes.toBytes(ele.day_course),
        Bytes.toBytes(cf),
        Bytes.toBytes(qualifier),
        ele.click_count)
    }
  }

  /**
   * 根据 RowKey 查询值
   */
  def count(day_course: String): Long = {
    val table = HBaseUtils.getInstance().getTable(tableName)

    val get = new Get(Bytes.toBytes(day_course))
    val value = table.get(get).getValue(cf.getBytes, qualifier.getBytes)

    if (value == null) {
      0L
    } else {
      Bytes.toLong(value)
    }
  }
}

File: SparkStreamingAction/spark-streaming-web/src/main/java/com/yourtion/bigdata/spark/project/web/dao/CourseClickCountDAO.java
package com.yourtion.bigdata.spark.project.web.dao;

import com.yourtion.bigdata.spark.project.web.domain.CourseClickCount;
import com.yourtion.bigdata.spark.project.web.utils.HBaseUtils;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.yourtion.bigdata.spark.project.web.utils.HBaseConfig.HBASE_CLICK_TABLE;

/**
 * 实战课程访问数量数据访问层
 *
 * @author yourtion
 */
@Component
public class CourseClickCountDAO {
    public static void main(String[] args) throws Exception {
        CourseClickCountDAO dao = new CourseClickCountDAO();
        List<CourseClickCount> list = dao.query("20200314");
        for (CourseClickCount model : list) {
            System.out.println(model.getName() + " : " + model.getValue());
        }
    }

    /**
     * 根据天查询
     */
    public List<CourseClickCount> query(String day) throws Exception {

        List<CourseClickCount> list = new ArrayList<>();

        // 去HBase表中根据day获取实战课程对应的访问量
        Map<String, Long> map = HBaseUtils.getInstance().query(HBASE_CLICK_TABLE, day);

        for (Map.Entry<String, Long> entry : map.entrySet()) {
            CourseClickCount model = new CourseClickCount();
            model.setName(entry.getKey());
            model.setValue(entry.getValue());

            list.add(model);
        }

        return list;
    }
}
